plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,80 +4,349 @@
+    let newNum = (n P.* d') P.- (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeSemigroup Rational where
+  {-# INLINEABLE (*) #-}
+  Rational n d * Rational n' d' =
+    let newNum = n P.* n'
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeMonoid Rational where
+  {-# INLINEABLE one #-}
+  one = Rational P.one P.one
+
+instance P.Module Integer Rational where
+  {-# INLINEABLE scale #-}
+  scale i (Rational n d) = let newNum = i P.* n
+                               gcd' = euclid newNum d in
+    Rational (newNum `Builtins.quotientInteger` gcd')
+             (d `Builtins.quotientInteger` gcd')
+
+instance P.ToData Rational where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (Rational n d) = P.toBuiltinData (n, d)
+
+-- These instances ensure that the following invariants don't break:
+--
+-- 1. The denominator is greater than 0; and
+-- 2. The numerator and denominator are coprime.
+--
+-- For invariant 1, fromBuiltinData yields Nothing on violation, while
+-- unsafeFromData calls error. Invariant 2 is kept maintained by use of
+-- unsafeRatio.
+
+instance P.FromData Rational where
+  {-# INLINEABLE fromBuiltinData #-}
+  fromBuiltinData dat = case P.fromBuiltinData dat of
+    P.Nothing -> P.Nothing
+    P.Just (n, d) -> if d P.== P.zero
+                     then Builtins.error ()
+                     else P.Just (unsafeRatio n d)
+
+instance P.UnsafeFromData Rational where
+  {-# INLINEABLE unsafeFromBuiltinData #-}
+  unsafeFromBuiltinData dat = case P.unsafeFromBuiltinData dat of
+    (n, d) -> if d P.== P.zero
+              then Builtins.error ()
+              else unsafeRatio n d
+
+-- | This mimics the behaviour of Aeson's instance for 'GHC.Rational'.
+instance ToJSON Rational where
+  toJSON (Rational n d) =
+    object
+      [ ("numerator", toJSON n)
+      , ("denominator", toJSON d)
+      ]
+
+-- | This mimics the behaviour of Aeson's instance for 'GHC.Rational'.
+instance FromJSON Rational where
+  parseJSON = withObject "Rational" Haskell.$ \obj -> do
+    n <- obj .: "numerator"
+    d <- obj .: "denominator"
+    case ratio n d of
+      Haskell.Nothing -> Haskell.fail "Zero denominator is invalid."
+      Haskell.Just r  -> Haskell.pure r
+
+-- | Makes a 'Rational' from a numerator and a denominator.
+--
+-- = Important note
+--
+-- If given a zero denominator, this function will error. If you don't mind a
+-- size increase, and care about safety, use 'ratio' instead.
+{-# INLINEABLE unsafeRatio #-}
+unsafeRatio :: Integer -> Integer -> Rational
+unsafeRatio n d
+  | d P.== P.zero = Builtins.error ()
+  | d P.< P.zero = unsafeRatio (P.negate n) (P.negate d)
```

Comment:
I think that `euclid a b` returns a number with the same absolute value as `gcd a b`, except that it's <=0 if b<0 and >= 0 if b>0. I think that means that we could miss out the second case because the third case would still do the right thing if `d < zero`.  That's rather subtle though, and if we were going to rely on that we'd need to be sure that it's correct (and have careful tests).

Reply:
This could potentially reduce code size, but in my tests, `unsafeRatio` is fairly small already. While I have [extensive test coverage](https://github.com/Liqwid-Labs/plutus-extra/tree/master/plutus-numeric/test) which could be ported to ensure this doesn't break, this is a fairly sizeable task. I'm happy to undertake it, but this PR is already nearly a month old.

Reply:
Yes, I agree.  If there was going to be a big win in terms of code size it might be worth making that change, but it would make the code rather fragile. I don't have a strong opinion either for or against.

Reply:
If the code is going to live here, the tests should be here too! I'm okay with that happening later, though.